src/Microsoft.OData.Core/UriParser/Binders/RangeVariableBinder.cs

```diff
@@ -24,7 +24,11 @@ internal static SingleValueNode BindRangeVariableToken(RangeVariableToken rangeV
         {
             ExceptionUtils.CheckArgumentNotNull(rangeVariableToken, "rangeVariableToken");
 
-            RangeVariable rangeVariable = state.RangeVariables.SingleOrDefault(p => p.Name == rangeVariableToken.Name);
+            // We are using FirstOrDefault instead of SingleOrDefault since we can have >1 $it rangeVariable in the Stack.
+            // If we added another $it rangeVariable to the Stack (when binding the expandToken or selectToken),
+            // We create the RangeVariableReferenceNode using the added $it rangeVariable.
+            // ASSUMPTION: Max 2 $it rangeVariables can be added to the Stack.
+            RangeVariable rangeVariable = state.RangeVariables.FirstOrDefault(p => p.Name == rangeVariableToken.Name);

```

Comment:
>FirstOrDefault [](start = 63, length = 14)

I am also confused by this logic.  If there are two $it range variables, how do we know that the desired range variable always the first?

Reply:
@mikepizzo  The first implicit rangeVariable is pushed to the stack in line 877. The one we are adding is pushed in line 889. The one on top of the stack will be the desired one.